model/pipeline/src/main/resources/org/apache/beam/model/pipeline/pipeline.json

```diff
@@ -0,0 +1,320 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+
+  "type": "object",
+
+  "$ref": "#/definitions/Pipeline",
+
+  "definitions": {
+    // A Pipeline contains the entire hierarchical graph including
+    // nodes for all values.
+    //
+    // Nodes (transform or value or other) are identified by some unique id. This need
+    // not be readable. It is simplest to think of it as a reified pointer or UUID.
+    // Cyclic pointer structures are forbidden throughout.
+    "Pipeline": {
+      "type": "object",
+      "additionalProperties": false,
+      "required": [ "userFns", "transforms", "values", "windowingStrategies", "coders" ],
+      "properties": {
+
+        // A mapping containing all user-definable functions in the pipeline.
+        // The keys must be unique to this pipeline definition, but otherwise are
+        // arbitrary.
+        "userFns": {
```

Comment:
Are these object ids used anywhere aside from cross referencing things in the pipeline description?

If that is the case, we could use an array and use array index to reference them, It will be a more compact pipeline, and remove some arbitrary ids. This will be less error prone at pipeline generation time (in regards to uniqueness requirement.)